### ISSUE TITLE
wrangler login just get stuck `Waiting for API token...`

### ISSUE THREAD

**Issue by rib** [cloudflare/wrangler-legacy#1703]
### 🐛 Bug Report

I found this closed issue https://github.com/cloudflare/wrangler/issues/1514 has a number of people reporting this same issue with the tool hanging, but looking at that original issue which was closed in October it looks like was for a different problem.

#### Environment

* operating system: `Ubuntu 20.04.1 LTS`
* output of `rustc -V`: `rustc 1.47.0 (18bf6b4f0 2020-10-07)`
* output of `node -v`: `v10.19.0`
* output of `wrangler -V`: `wrangler 1.12.3`
* contents of `wrangler.toml`
```
name = "test"
type = "javascript"
account_id = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
workers_dev = true
route = ""
zone_id = ""
```

(Simply created via `wrangler generate test`)

#### Steps to reproduce

run `wrangler login`

#### What did you expect to see?

After authorizing Cloudflare via the opened web browser, after the token is created, I expected `wrangler login` to continue.

#### What did you see instead?

I do see the browser open (Firefox) and I'm able to authorize the creation of a new token which does happen successfully but the `wrangler login` tool remains stuck forever with `Waiting for API token...`

Since it was suggested in #1514 that it might make a difference to delete existing tokens first I also tried deleting all tokens but that didn't seem to make a difference for me.

**Comment by siba2893** [comment]
This happens to me too. Can use ```wrangle login``` cause it gets stucked. :/

**Comment by nolandubeau** [comment]
I am experiencing the same issue.  I wonder if it has to do with the fact that I have two cloudflare accounts.  When I do `wrangler whoami`, I see the old account that I do not want to use.  It could be useful to have a `wrangler logout` command which other CLI's (like Firease-CLI) have.  At first I thought this may have been an issue with the api token callback back to VSCode, but the error is also happening in Terminal (on MacOS).

**Comment by DariuszOstolski** [comment]
I have the same issue, I have one cloudflare account. I'm using Ubuntu with Firefox as default browser and containers extension.

**Comment by nolandubeau** [comment]
I was able to get the login and auth callback from the browser to work just now.  Not sure if this works for anyone else, but I ran the command `wrangler config --api-key`, which is the command to set the global api key, and which is not recommended.  I cancelled out from that command and then ran `wrangler login` again, and it authenticated successfully.  Could have been chance, but maybe that might work for someone else as the command may have cleared something from the config.

**Comment by danbars** [comment]
Same issue here too. 
Chrome 87 on Mac 10.15.5

**Comment by ClementParis016** [comment]
I just encountered the same issue, weirdly enough @nolandubeau's workaround worked for me too (thanks ❤️ )

**Comment by lemaguilera** [comment]
> I was able to get the login and auth callback from the browser to work just now. Not sure if this works for anyone else, but I ran the command `wrangler config --api-key`, which is the command to set the global api key, and which is not recommended. I cancelled out from that command and then ran `wrangler login` again, and it authenticated successfully. Could have been chance, but maybe that might work for someone else as the command may have cleared something from the config.

It works! thanks

**Comment by lougayou** [comment]
I ran into the same issue:
Ubuntu 20.04
Chrome 88.0.4324.96
wrangler 1.13.0
node v14.4.0

Thanks @nolandubeau , your workaround fixed the issue!

**Comment by Daggy1234** [comment]
Same issue on macos

**Comment by yucongo** [comment]
Sometimes if you already logged in and tried to log in again, you'll get stuck. In any case, that was what happen to me with `wrangler 1.13.0` in `ubuntu 19`. 

I got stuck with the same message `Waiting for API token...` and tried many methods suggsested from the net. Nothing worked. It turned out I already logged in last time. 

To find out whether you already logged in, do
```wrangler whoami```
If your account name and id show up, you are good to go.

**Comment by omarabid** [comment]
Same issue for freshly installed wrangler (using Cargo). Version 1.13.0. I'm using Archlinux (Kernel version 5.10.11)

**Comment by tongyifan** [comment]
Same issue. I fixed it by fill API token manually.
1. Use `wrangler login` to open login link.
2. Press F12, then switch it to network tab.
3. Press "Authorize Wrangler", then wait it success.
4. Check the lastest request, find a request like `https://dash.cloudflare.com/api/v4/user/tokens/SOME_HEX_STRING_HERE/value` , then find your API token in the response.
5. Run command `wrangler config`, paste the API token into it and press Enter, then you will get "Successfully configured. You can find your configuration file at: ..."

**Comment by xortive** [comment]
Hey folks, are y'all reaching a page that looks like:
![Screenshot from 2021-03-02 09-38-42](https://user-images.githubusercontent.com/566737/109672916-1f19bf00-7b3b-11eb-93db-9eff3575f9d0.png)

and after clicking "Authorize Wrangler", wrangler login is still stuck?

If you aren't reaching this page, try logging in and opening the link that `wrangler login` prints to your console.

If you are clicking "Authorize Wrangler" and `wrangler login` is still getting stuck, I'd like to get more information to reproduce this problem.

**Comment by danbars** [comment]
Hi @xortive , I reach this page, but after clicking the authorize button, the cli is still stuck.
Happens consistently on all worker projects that I have, workers and worker-sites.
Here's one toml for example

```toml
name = "form-data-app"
type = "webpack"
account_id = "c4axxxxxxxxxx710"
workers_dev = false
route = "https://app.form-data.com/*"
zone_id = "f6bxxxxxxxxxxxxxxxxe15"

[site]
bucket = "./dist/pwa"
entry-point = "workers-site"
```

Happens on Chrome 88 on Mac 10.15.5, but Also happened on previous versions of both mac and chrome.
Let me know what other details do you need.

**Comment by samarpanda** [comment]
> Hey folks, are y'all reaching a page that looks like:
> ![Screenshot from 2021-03-02 09-38-42](https://user-images.githubusercontent.com/566737/109672916-1f19bf00-7b3b-11eb-93db-9eff3575f9d0.png)
> 
> and after clicking "Authorize Wrangler", wrangler login is still stuck?
> 
> If you aren't reaching this page, try logging in and opening the link that `wrangler login` prints to your console.
> 
> If you are clicking "Authorize Wrangler" and `wrangler login` is still getting stuck, I'd like to get more information to reproduce this problem.

Yes, after clicking the wrangler login it is stuck. Let me know what info do you need. I can share.

**Comment by Daggy1234** [comment]
After clicking it got stuck, I was able to fix it by manually setting it a long time ago. On macos with homebrew.

**Comment by Froxz** [comment]
Have same issue,

After calling `wrangler config` added token manually.

Calling `wrangler whoami` outputs correct email and token.

But when runing `wrangler dev` - shows `You have not provided your Cloudflare credentials.` Anynoe had same issue?

**Comment by churchthecat** [comment]
Managed to login , by 
wrangler login ->authorize
selecting roll on the API Token page for wrangler
copy token 
wrangler config ->paste token

**Comment by antholeole** [comment]
Extra info - may be browser dependent? I attempted three times in Safari, both times I got stuck on "waiting for API token..." switched my default browser to Firefox, tried again, and it worked fine. 

May be due to other external factors and just so happened that it worked 1 / 4 times, but might be an avenue worth investigating.

**Comment by intellix** [comment]
Was trying this from VSCode Integrated Terminal to Chrome and it was stuck for me. I tried it in OSX Terminal to Chrome and then it worked as expected

**Comment by fawazahmed0** [comment]
same issue here

**Comment by fawazahmed0** [comment]
> 
> 
> Same issue. I fixed it by fill API token manually.
> 
>     1. Use `wrangler login` to open login link.
> 
>     2. Press F12, then switch it to network tab.
> 
>     3. Press "Authorize Wrangler", then wait it success.
> 
>     4. Check the lastest request, find a request like `https://dash.cloudflare.com/api/v4/user/tokens/SOME_HEX_STRING_HERE/value` , then find your API token in the response.
> 
>     5. Run command `wrangler config`, paste the API token into it and press Enter, then you will get "Successfully configured. You can find your configuration file at: ..."

Thanks, this worked

**Comment by roelandmoors** [comment]
After the token was created I just aborted the login command. Then find the token [here](https://dash.cloudflare.com/profile/api-tokens). Next to the token is a button with 3 dots. Click and select 'roll'.
Now you get a new token that you can use with `wrangler config`

**Comment by timhaak** [comment]
From my side, it worked once I created the following folder.

mkdir -p ~/.wrangler/config/

**Comment by jide** [comment]
Same issue using VSCode and Brave. Kind of frustrating when you're onboarding ;)

**Comment by RemiKalbe** [comment]
Same issue on macos 11.2.3 with brave.

**Comment by nilslice** [comment]
Hey all - we're currently looking into this issue & have some questions if anyone can share details on the following:

1. when `wrangler login` remained stuck waiting on the token, was this your first time logging in via CLI? 
2. if you got the "Wrangler is now authorized with a new API Token" screen but still observed the CLI to hang, can you check if you ended up getting anything written to `~/.wrangler/config/default.toml` or the equivalent path on your OS?


Thank you -- lots of weird little edge cases here, so all details are really helpful.

**Comment by nilslice** [comment]
@timhaak, re: 
> From my side, it worked once I created the following folder.
> mkdir -p ~/.wrangler/config/

Did the CLI show any permission error indicating that it couldn't write to disk or create any files? 

And to confirm, you created that directory, and then re-executed `wrangler login`, which worked as expected?

**Comment by Yrlish** [comment]
Hi @nilslice!

Just because you wanted me to test, wrangler doesn't seem to get stuck as described in #1883. Although, the token got replaced so it invalidated one environment when logging into another.

1. 1.16.1
2. Yes
3. -

**Comment by nilslice** [comment]
@Yrlish - thank you!

> Just because you wanted me to test, wrangler doesn't seem to get stuck as described in #1883. 

😄  haha that seems to be the way these things work! 

So both devices are now able to use `wrangler login` ? The device change should invalidate the other, but it shouldn't hang when trying to revalidate on the second device. Are you able to get the second device to auth as expected, or does it still hang?

**Comment by Yrlish** [comment]
> 😄 haha that seems to be the way these things work!

haha yes indeed, very annoying 😛 

Yes, both environments logged in using `wrangler login` without any problems when I just tested. First environment created a token and received it successfully, second environment replaced the token and received it successfully.

**Comment by JoyceBabu** [comment]
@nilslice I am also facing this issue. Please find the requested details below. 

> 1. which version of wrangler are you running? `wrangler --version`

```👷 ✨  wrangler 1.16.1```

> 2. when `wrangler login` remained stuck waiting on the token, was this your first time logging in via CLI?

First time.

> 3. if you got the "Wrangler is now authorized with a new API Token" screen but still observed the CLI to hang, can you check if you ended up getting anything written to `~/.wrangler/config/default.toml` or the equivalent path on your OS?

Even the directory was not created. I tried creating the directory as @timhaak and retried `wrangler login`, but it still did not work.

> **Additionally**: if you can retry the login flow with logging enabled, and share the output (removing any tokens if they are visible), please do by running:
> 
> `RUST_LOG=debug wrangler login`
> 
> Thank you -- lots of weird little edge cases here, so all details are really helpful.

https://pastebin.com/1D86d5Nc

**Comment by victorbjorklund** [comment]
I still have this issue using wrangler 1.17.0 on a Mac. Nothing gets written to ~/.wrangler/config/. Terminal just stuck at "Waiting for API token..." forever while website says "Wrangler is now authorized with a new API Token. You can now close this page."

**Comment by matt-allan** [comment]
I'm seeing this too. Tested with yarn, npm and cargo wrangler installs and Safari and Firefox as the default browser. MacOS Big Sur 11.4 (20F71).

**Comment by alanshaw** [comment]
I had the same issue using Firefox as default browser. I copied the link into Chrome and then was able to login.

**Comment by leoplct** [comment]
I've run mkdir -p ~/.wrangler/config/ but I got stuck at "Waiting for API token...".

**Comment by khromov** [comment]
My problem was that I tried running `wrangler` through `npx`, so for example `npx @cloudflare/wrangler login` **does not work**.

**Comment by nilslice** [comment]
@khromov - thanks for reporting. did you also experience wrangler getting stuck at "Waiting for API token...", or did you get different output?

**Comment by khromov** [comment]
@nilslice Yes, it was stuck at "Waiting for API token...". 

I was running the command from a VSCode prompt using Windows 10 and Ubuntu on WSL2. The browser did not open automatically, I had to paste the URL manually. 

(The browser did not open automatically when installed via `-g` either, but logging in worked fine.)

**Comment by nneil** [comment]
Same problem under Windows 10. The window opens and the token is generated, but the `Waiting for API token...` never returns. I generated an API token manually and added it with `wrangle config`. It worked well enough to go through the quickstart.

**Comment by JIBSIL** [comment]
My problem was that there were too many API tokens (or maybe the name was conflicting?). As soon as I removed all of my API tokens that I wasn't using with the same name, it cleared up.

**Comment by adammonago** [comment]
Same issue here. using chrome on osx.  @nolandubeau workaround fixed the issue for me as well!

**Comment by neerolyte** [comment]
```
$ wrangler --version
wrangler 1.17.0
```

This is my first time attempting to login via the CLI.

```
$ ls -la ~/.wrangler/config/
total 8
drwxrwxr-x 2 dschoen dschoen 4096 Aug 10 14:39 .
drwxrwxr-x 5 dschoen dschoen 4096 Aug 10 14:38 ..
```

```
$ RUST_LOG=debug wrangler login
[2021-08-10T04:42:49Z INFO  wrangler::settings::global_config] No $WRANGLER_HOME detected, using $HOME
Allow Wrangler to open a page in your browser? [y/n]
y
💁  Opened a link in your default browser: https://dash.cloudflare.com/wrangler?key=...
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::producer_ready; state=State[count=0; consuming=false; producing=false; lifecycle=New]
[2021-08-10T04:42:51Z DEBUG eventual::core]   - transitioned from State[count=0; consuming=false; producing=false; lifecycle=New] to State[count=0; consuming=false; producing=false; lifecycle=ProducerWait]
⠁   Waiting for API token...
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::consumer_await
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::consumer_ready; state=State[count=0; consuming=false; producing=false; lifecycle=ProducerWait]
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::consumer_wait; state=State[count=0; consuming=false; producing=false; lifecycle=ProducerWait]
[2021-08-10T04:42:51Z DEBUG eventual::core]   - transitioned from State[count=0; consuming=false; producing=false; lifecycle=ProducerWait] to State[count=1; consuming=false; producing=false; lifecycle=ConsumerWait]
[2021-08-10T04:42:51Z DEBUG eventual::core]   - notifying producer
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::notify_producer
[2021-08-10T04:42:51Z DEBUG eventual::core]   - transitioned from State[count=1; consuming=false; producing=false; lifecycle=ConsumerWait] to State[count=1; consuming=false; producing=true; lifecycle=ConsumerWait]
[2021-08-10T04:42:51Z DEBUG eventual::core]   - Invoking producer; state=State[count=1; consuming=false; producing=true; lifecycle=ConsumerWait]
[2021-08-10T04:42:51Z DEBUG eventual::stream] Sender::poll; is_ready=true
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::producer_poll; state=State[count=1; consuming=false; producing=true; lifecycle=ConsumerWait]
[2021-08-10T04:42:51Z DEBUG syncbox::thread_pool] running task; tp-state=Running; worker_count=0
[2021-08-10T04:42:51Z DEBUG syncbox::thread_pool] spawning new worker thread
[2021-08-10T04:42:51Z DEBUG syncbox::thread_pool] task submitted to queue
[2021-08-10T04:42:51Z DEBUG eventual::core]   - Producer invoked; state=State[count=1; consuming=false; producing=false; lifecycle=ConsumerWait]
[2021-08-10T04:42:51Z DEBUG eventual::core] Core::consumer_await - parking thread
⠒   Waiting for API token...
[2021-08-10T04:42:52Z DEBUG syncbox::thread_pool] worker processing task
[2021-08-10T04:42:52Z DEBUG syncbox::thread_pool] ~~ running scheduled task
[2021-08-10T04:42:52Z DEBUG eventual::core] Core::complete; state=State[count=1; consuming=false; producing=false; lifecycle=ConsumerWait]; success=true; last=false
[2021-08-10T04:42:52Z DEBUG eventual::core]   - transitioned from State[count=1; consuming=false; producing=false; lifecycle=ConsumerWait] to State[count=1; consuming=false; producing=false; lifecycle=Ready]
[2021-08-10T04:42:52Z DEBUG eventual::core]   - notifying consumer
[2021-08-10T04:42:52Z DEBUG eventual::core] Core::consumer_await - unparking thread
[2021-08-10T04:42:52Z DEBUG eventual::core]   - consumer notified
[2021-08-10T04:42:52Z DEBUG eventual::core] Core::producer_ready; state=State[count=1; consuming=false; producing=false; lifecycle=Ready]
[2021-08-10T04:42:52Z DEBUG eventual::core]   - transitioned from State[count=1; consuming=false; producing=false; lifecycle=Ready] to State[count=1; consuming=false; producing=false; lifecycle=ReadyProducerWait]
[2021-08-10T04:42:52Z DEBUG syncbox::thread_pool] worker waiting for task
[2021-08-10T04:42:52Z DEBUG eventual::core] Core::consumer_poll; state=State[count=1; consuming=true; producing=false; lifecycle=Ready]
[2021-08-10T04:42:52Z DEBUG eventual::core]   - transitioned from State[count=1; consuming=false; producing=false; lifecycle=ReadyProducerWait] to State[count=1; consuming=false; producing=false; lifecycle=ProducerWait] (consuming value)
⠋   Waiting for API token...
[2021-08-10T04:43:22Z DEBUG eventual::core] Core::cancel; state=State[count=1; consuming=false; producing=false; lifecycle=ProducerWait]
[2021-08-10T04:43:22Z DEBUG eventual::core]   - transitioned from State[count=1; consuming=false; producing=false; lifecycle=ProducerWait] to State[count=1; consuming=false; producing=false; lifecycle=Canceled]
[2021-08-10T04:43:22Z DEBUG eventual::core] Core::notify_producer
[2021-08-10T04:43:22Z DEBUG eventual::core]   - transitioned from State[count=1; consuming=false; producing=false; lifecycle=Canceled] to State[count=1; consuming=false; producing=true; lifecycle=Canceled]
[2021-08-10T04:43:22Z DEBUG eventual::core]   - Invoking producer; state=State[count=1; consuming=false; producing=true; lifecycle=Canceled]
[2021-08-10T04:43:22Z DEBUG eventual::stream] Sender::poll; is_ready=false
[2021-08-10T04:43:22Z DEBUG eventual::core] Core::producer_poll; state=State[count=1; consuming=false; producing=true; lifecycle=Canceled]
[2021-08-10T04:43:22Z DEBUG syncbox::thread_pool] shutdown; tp-state=Running; target=Shutdown; worker_count=1
[2021-08-10T04:43:22Z DEBUG syncbox::thread_pool]   enqueuing no-ops; count=1
[2021-08-10T04:43:22Z DEBUG syncbox::thread_pool] received no-op token -- shutting down
Error: error sending request for url (https://api.cloudflare.com/client/v4/workers/token): operation timed out

Caused by:
    operation timed out
```

Wrangler is _not_ actually opening the link in any browser I have been able to find, but if I click the link presented in the terminal I end up getting to the `Wrangler is now authorized with a new API Token\nYou can now close this page.` prior to the timeout occuring in the CLI.

**Comment by vishal-1408** [comment]
I faced the same issue in my laptop running window10. I found a simple workaround for this issue. We can create custom tokens in the Cloudflare account with the settings same as the generated token. Enter this token manually using the `wrangler config `command and you are good to go.

**Comment by zaerl** [comment]
I solved this issue removing the `.wrangler/` folder from the home directory and trying again. The CLI will generate a new `default.toml` file.

**Comment by famzah** [comment]
I had to delete the two existing tokens in the Cloudflare Dashboard, in order to fix this.

**Comment by brandiqa** [comment]
Using the latest version of Wrangler 1.19.2 in Zorin OS(Ubuntu distro) and `wrangler login` doesn't work as expected. Had to resort to the manual method.

**Comment by falcon03** [comment]
Another user facing this issue: wrangler login does not work. Tested with wrangler 2.0.14...

Checking from the cloudflare dashboard, it looks like wrangler login does not generate an api token at all!

**Comment by fhirfly** [comment]
my wrangler on Ubuntu hangs no matter if its login or publish.   login gets a token (writes to .wrangler/config) but then hangs.  same thing when I run publish, just hangs.  No way to kill it.

**Comment by mythz** [comment]
The docs for using curl in [Using wrangler login on a remote machine](https://developers.cloudflare.com/workers/wrangler/commands/#using-wrangler-login-on-a-remote-machine) didn't work for me, what did was opening port forwarding to the remote server which lets me follow their OAuth in my desktop browser and successfully redirect back to wrangler login running on my remote server:

```bash
ssh -L 8976:0.0.0.0:8976 -N name@server
```

**Comment by IgnisDa** [comment]
I had this problem in my devcontainer where I could not forward port `8976` for some reason. Here is what worked for me:

1. Run `wrangler login`
2. Go through all the steps until you get to the page which tries to connect to `http://localhost:8976/oauth/...`
3. Run `docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' container_name` to get the IP address of the running container
4. Replace `localhost` in the browser URL you got in step 2 with the IP address you got in step 3 and then reload the page.

**Comment by JacobMGEvans** [comment]
Thanks for creating this issue! Wrangler v1 is now deprecated and support is only being provided for critical updates or security concerns. As such, we are closing this PR.

New versions of Wrangler are maintained in the [workers-sdk repo](https://github.com/cloudflare/workers-sdk). If you are running into a similar issue with wrangler v2, please report it or create a new PR in the [workers-sdk repo](https://github.com/cloudflare/workers-sdk). For more info about wrangler v1 deprecation, please check out our [blog post](https://blog.cloudflare.com/deprecating-wrangler-v1).
